src/Agent.Listener/LocalRunner.cs

```diff
@@ -56,6 +90,422 @@ public sealed class LocalRunner : AgentService, ILocalRunner
+      {JsonConvert.ToString(variable.Key ?? string.Empty)}: {JsonConvert.ToString(variable.Value ?? string.Empty)}");
+                }
+
+            builder.Append($@"
+    }},
+    ""systemConnection"": {{
+      ""data"": {{
+        ""ServerId"": ""00000000-0000-0000-0000-000000000000"",
+        ""ServerName"": ""127.0.0.1""
+      }},
+      ""name"": ""SystemVssConnection"",
+      ""url"": ""https://127.0.0.1/vsts-agent-local-runner"",
+      ""authorization"": {{
+        ""parameters"": {{
+          ""AccessToken"": ""dummy-access-token""
+        }},
+        ""scheme"": ""OAuth""
+      }},
+      ""isReady"": false
+    }}
+  }}
+}}");
+                string message = builder.ToString();
+                try
+                {
+                    jobs.Add(JsonUtility.FromString<AgentJobRequestMessage>(message));
+                }
+                catch
+                {
+                    Dump("Job message JSON", message);
+                    throw;
+                }
+            }
+
+            return jobs;
+        }
+
+        private async Task<TaskDefinition> GetDefinitionAsync(TaskStep task, CancellationToken token)
+        {
+            var available = await GetAvailableTasksAsync(token);
+            ArgUtil.NotNull(task.Reference, nameof(task.Reference));
+            ArgUtil.NotNullOrEmpty(task.Reference.Name, nameof(task.Reference.Name));
+            List<TaskDefinition> definitions;
+            if (!available.TryGetValue(task.Reference.Name, out definitions))
+            {
+                throw new Exception($"Unable to resolve task {task.Reference.Name}");
+            }
+
+            // Attempt to find an exact match.
+            TaskDefinition match = definitions.FirstOrDefault(definition => string.Equals(GetVersion(definition).ToString(), task.Reference.Version ?? string.Empty, StringComparison.Ordinal));
+
+            // Attempt to find a match from the server.
+            if (match == null && _httpClient != null && Regex.IsMatch(task.Reference.Version ?? string.Empty, @"^[0-9]+\.[0-9]+\.[0-9]+$"))
+            {
+                string cacheKey = $"{task.Reference.Name}@{task.Reference.Version}";
+                if (!_queryCache.TryGetValue(cacheKey, out match))
+                {
+                    _term.WriteLine($"Querying server for task name='{task.Reference.Name}', version='{task.Reference.Version}', id='{definitions[0].Id}'.");
+                    match = await _httpClient.GetTaskDefinitionAsync(taskId: definitions[0].Id, versionString: task.Reference.Version);
+                    _queryCache.Add(cacheKey, match);
+                }
+            }
+
+            // Attempt to find the best match from the "available" cache.
+            if (match == null)
+            {
+                var versionPattern = new StringBuilder("^");
+                foreach (char c in (!string.IsNullOrEmpty(task.Reference.Version) ? task.Reference.Version : ".*"))
```

Comment:
make this more clear: empty version = latest, otherwise convert to regex (e.g. `1.*` =\> `^1\..*$`)